module/zfs/dbuf.c

```diff
@@ -2029,11 +2030,14 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
 			    FTAG);
 		}
 	}
+#endif
 
 	if (tx->tx_txg > dn->dn_dirty_txg)
 		dn->dn_dirty_txg = tx->tx_txg;
 	mutex_exit(&dn->dn_mtx);
 
+	dnode_set_dirtyctx(dn, tx, db);
```

Comment:
It looks like you've accidentally kept both versions of the `dn->dn_dirtyctx == DN_UNDIRTIED)` block.  The one above wrapped in `ZFS_DEBUG` and this call to `dnode_set_dirtyctx`.  The one above also incorrectly uses `kmem_alloc` when updating the tag.

Since we need to take the `&dn->dn_mtx` anyway to update `dn->dn_dirty_txg`.  There's probably not a significant benefit is wrapping all the `dn_dirtyctx_*` fields with `ZFS_DEBUG`.  So why don't we keep the change minimal:

```c
       mutex_enter(&dn->dn_mtx);
       dnode_set_dirtyctx(dn, tx, db);

	if (tx->tx_txg > dn->dn_dirty_txg)
		dn->dn_dirty_txg = tx->tx_txg;

        mutex_exit(&dn->dn_mtx);
```